.github/workflows/main.yml

```diff
@@ -0,0 +1,74 @@
+name: CI
+
+# Controls when the action will run. Triggers the workflow on push or pull request
+# events but only for the master branch
+on:
+  push:
+    branches: [ master ]
+  pull_request:
+    branches: [ master ]
+
+jobs:
+  build-toolchain:
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [ubuntu-16.04, ubuntu-18.04, ubuntu-20.04]
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+        # we don't need submodules for building the toolchain
+
+      # toolchain takes ~40min to build, so we want to avoid that whenever possible!
+      - uses: actions/cache@v2
+        id: toolchain-cache
+        with:
+          path: toolchain.tar.gz
+          # gcc/binutils/etc versions are in build_toolchain.sh, so that'll be part of our cache key
+          key: ${{ matrix.os }}-${{ hashFiles('ndless-sdk/toolchain/build_toolchain.sh') }}
+      - name: Install dependencies
+        if: steps.toolchain-cache.outputs.cache-hit != 'true'
+        run: sudo apt install -y libmpfr-dev libmpc-dev
+      - name: Build
+        if: steps.toolchain-cache.outputs.cache-hit != 'true'
+        working-directory: ndless-sdk/toolchain
+        run: ./build_toolchain.sh
+      - name: Package toolchain
+        if: steps.toolchain-cache.outputs.cache-hit != 'true'
+        run: tar -cvzf toolchain.tar.gz ndless-sdk/toolchain/install
+      
+      - name: Upload built toolchain
+        uses: actions/upload-artifact@v2
+        with:
+          name: toolchain-${{ matrix.os }}
+          path: toolchain.tar.gz
+  build-sdk:
+    needs: build-toolchain
+    
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [ubuntu-16.04, ubuntu-18.04, ubuntu-20.04]
+    
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+        with:
+          submodules: 'recursive'
+          fetch-depth: 0 # submodule checkout does not work with depth=1 (the default)
+      
+      - name: Download built toolchain
+        uses: actions/download-artifact@v2
+        with:
+          name: toolchain-${{ matrix.os }}
+      - name: Extract toolchain
+        run: tar -xvzf toolchain.tar.gz ndless-sdk/toolchain/install
+      - name: Set up PATH
+        run: |
+          echo "::add-path::$GITHUB_WORKSPACE/ndless-sdk/toolchain/install/bin"
+          echo "::add-path::$GITHUB_WORKSPACE/ndless-sdk/bin"
+      - name: Install dependencies
+        run: sudo apt install -y libboost-program-options-dev libmpfr-dev libmpc-dev
```

Comment:
At least libmpfr-dev seems to be required: https://github.com/shinyblink/Ndless/runs/945668803?check_suite_focus=true

If you can't see the detailed logs, here's the relevant excerpt from `build-sdk` on ubuntu-16.04:

```
make -C ndless-sdk
make[1]: Entering directory '/home/runner/work/Ndless/Ndless/ndless-sdk'
make -C libsyscalls
make[2]: Entering directory '/home/runner/work/Ndless/Ndless/ndless-sdk/libsyscalls'
arm-none-eabi-g++ -mcpu=arm926ej-s -std=c++11 -nostdlib -O3 -fPIE -fno-exceptions -fno-rtti -mlong-calls -Wall -Werror -I ../include/ -I ../thirdparty/nspire-io/include/ -D_TINSPIRE -ffunction-sections -fdata-sections -c stubs.cpp -o stubs.o
/home/runner/work/Ndless/Ndless/ndless-sdk/toolchain/install/libexec/gcc/arm-none-eabi/10.1.0/cc1plus: error while loading shared libraries: libmpfr.so.6: cannot open shared object file: No such file or directory
Makefile:18: recipe for target 'stubs.o' failed
make[2]: *** [stubs.o] Error 1
make[2]: Leaving directory '/home/runner/work/Ndless/Ndless/ndless-sdk/libsyscalls'
make[1]: *** [build-libsyscalls] Error 2
Makefile:14: recipe for target 'build-libsyscalls' failed
make: *** [build-ndless-sdk] Error 2
make[1]: Leaving directory '/home/runner/work/Ndless/Ndless/ndless-sdk'
Makefile:19: recipe for target 'build-ndless-sdk' failed
##[error]Process completed with exit code 2.
```

The non-dev packages might also work for this, but they're different for 16.04 (libmpfr4) and 18.04/20.04 (libmpfr6): https://packages.ubuntu.com/search?keywords=libmpfr

Reply:
That's because the toolchain binaries were built against libmpfr, but then copied into a fresh environment without installing deps. This is one of the reasons why I think that having this in a single job is cleaner.